crates/federate/src/federation_queue_state.rs

```diff
@@ -0,0 +1,52 @@
+use crate::util::ActivityId;
+use anyhow::Result;
+use chrono::{DateTime, TimeZone, Utc};
+use diesel::prelude::*;
+use diesel_async::RunQueryDsl;
+use lemmy_db_schema::utils::{get_conn, DbPool};
+
+#[derive(Queryable, Selectable, Insertable, AsChangeset, Clone)]
+#[diesel(table_name = lemmy_db_schema::schema::federation_queue_state)]
+#[diesel(check_for_backend(diesel::pg::Pg))]
+pub struct FederationQueueState {
+  /// domain of the instance (primary key)
+  pub domain: String,
+  pub last_successful_id: ActivityId, // todo: i64
+  pub fail_count: i32,
+  pub last_retry: DateTime<Utc>,
+}
+
+impl FederationQueueState {
+  /// load or return a default empty value
+  pub async fn load(pool: &mut DbPool<'_>, domain_: &str) -> Result<FederationQueueState> {
+    use lemmy_db_schema::schema::federation_queue_state::dsl::{domain, federation_queue_state};
+    let conn = &mut get_conn(pool).await?;
+    Ok(
+      federation_queue_state
+        .filter(domain.eq(&domain_))
+        .select(FederationQueueState::as_select())
+        .get_result(conn)
+        .await
+        .optional()?
+        .unwrap_or(FederationQueueState {
+          domain: domain_.to_owned(),
+          fail_count: 0,
+          last_retry: Utc.timestamp_nanos(0),
+          last_successful_id: 0, // todo: start at current id not from beginning
```

Comment:
Not sure if this todo is important.

Reply:
Mh, it kind of is. This value determines at which point in time the federation for a new instance starts. 

I've updated this value to be set to the newest activity at the point in time where the instance is first seen.